### PR TITLE
[clang] Warn on mismatched RequiresCapability attributes

### PR DESCRIPTION
**Pull request by tbaederr**
There's a problem with this attribute where the declaration in a header file doesn't have the attribute, but the definition in the source file has. As a result, the attribute doesn't take effect when just the header file is included.

This is a WIP patch to address this problem and emit a warning for it. I'm not sure if this is the best place to do it. As discussed with @AaronBallman on Discord, it can't be done in `Sema::MergeFunctionDecl()`, because this is a late parsed attribute and they aren't available yet when we merge the decl.

### CODE REVIEW

**Comment by AaronBallman** [comment]
> There's a problem with this attribute where the declaration in a header file doesn't have the attribute, but the definition in the source file has. As a result, the attribute doesn't take effect when just the header file is included.

Errr, that's the behavior of basically all inheritable attributes -- the design idea there is that the attribute is additive information that's only visible from the declaration with the attribute onward. So I think some users will appreciate this warning because they might not have been aware that the attribute was not really doing much for them, but I also think some users will be frustrated because they've done things like:
```
// Source.cpp
void func();

void foo() {
  func(); // Doesn't know about the attribute, but that's intentional for whatever reason.
}

ATTR void func();

void bar() {
  func();  // Does see the attribute
}

void func() {}

void baz() {
  func(); // Still sees the attribute
}
```
I could imagine this being useful for `RequiresCapability` in cases where you only want to require the capability internally as an implementation detail, but not make it part of the public interface of the function.

@aaronpuchert -- do you have thoughts here?

**Comment by aaronpuchert** [comment]
Yeah, it's a tricky question. On one hand there have been previous issues like #42194 (which this would basically address), and it would definitely improve consistency and prevent beginner's mistakes. On the other hand it seems useful to allow adding attributes e.g. to third-party libraries where you don't control the headers.

Perhaps we should introduce such checks, but put them under a subflag of `-Wthread-safety-attributes` to allow opting out of them (even locally via pragmas). And of course it should apply more or less to every thread safety annotation, not just `requires_capability`.

**Comment by tbaederr** [comment]
@AaronBallman Is the prototype of the implementation here at least sound? Or should this go somewhere completely different? Maybe in one of the TSA source files?

**Comment by AaronBallman** [comment]
> @AaronBallman Is the prototype of the implementation here at least sound? Or should this go somewhere completely different? Maybe in one of the TSA source files?

This is a semantic concern rather than a syntactic one, so I would expect this to happen in Sema rather than in Parser. My intuition is that this would be most natural in `mergeDeclAttributes()` as that's where this sort of thing typically lives for non-late parsed attributes. We merge the attributes when merging function declarations, so perhaps we need to add `mergeLateParsedDeclAttributes()` for ones that require late parsing?

CC @erichkeane in case he has opinions.

**Comment by tbaederr** [comment]
I've resorted to do this in `ThreadSafety.cpp` now, since all other places where we merge function definition and declaration happen before we have the late-parsed attributes available.

From a diagnostic POV, this seems quite complicated since we have N declarations and need to report differences in M attributes (and each one of them can have >=1 parameter...). For the POC implementation here, I've left if with just reporting a mismatch in attributes. For example with `test/SemaCXX/warn-thread-safety-analysis.cpp`:
```console
error: 'expected-warning' diagnostics seen but not expected:
  File /home/tbaeder/code/llvm-project/clang/test/SemaCXX/warn-thread-safety-analysis.cpp Line 2171: attribute mismatch between function definition and declaration of 'foo2'
  File /home/tbaeder/code/llvm-project/clang/test/SemaCXX/warn-thread-safety-analysis.cpp Line 2172: attribute mismatch between function definition and declaration of 'foo3'
  File /home/tbaeder/code/llvm-project/clang/test/SemaCXX/warn-thread-safety-analysis.cpp Line 2222: attribute mismatch between function definition and declaration of 'fooF2'
  File /home/tbaeder/code/llvm-project/clang/test/SemaCXX/warn-thread-safety-analysis.cpp Line 3182: attribute mismatch between function definition and declaration of 'foo1'
  File /home/tbaeder/code/llvm-project/clang/test/SemaCXX/warn-thread-safety-analysis.cpp Line 3183: attribute mismatch between function definition and declaration of 'foo2'
  File /home/tbaeder/code/llvm-project/clang/test/SemaCXX/warn-thread-safety-analysis.cpp Line 5220: attribute mismatch between function definition and declaration of 'test'
  File /home/tbaeder/code/llvm-project/clang/test/SemaCXX/warn-thread-safety-analysis.cpp Line 5221: attribute mismatch between function definition and declaration of 'test2'

```

**Comment by theuni** [comment]
I tried out this WIP on our codebase after we ran into a problem where an annotation was added to a definition rather than a declaration, leading to a (silently) missed capability check.

This addition seems to work fine, but it triggers warnings when annotations are added to both the definition and declaration. This can be seen here: https://github.com/bitcoin/bitcoin/pull/30316

I went ahead and PR'd the removals because they're straightforward enough, but it seems (to me) that those these are all false positives. If the same annotation is listed in the declaration and definition, should it really warn?

**Comment by tbaederr** [comment]
> I tried out this WIP on our codebase after we ran into a problem where an annotation was added to a definition rather than a declaration, leading to a (silently) missed capability check.
> 
> This addition seems to work fine, but it triggers warnings when annotations are added to both the definition and declaration. This can be seen here: [bitcoin/bitcoin#30316](https://github.com/bitcoin/bitcoin/pull/30316)
> 
> I went ahead and PR'd the removals because they're straightforward enough, but it seems (to me) that those these are all false positives. If the same annotation is listed in the declaration and definition, should it really warn?

No I don't think it should, at least that was not the intention of this patch. As it is right now, the code should warn for attributes listed in the definition, but _now_ in the declaration(s).

**Comment by aaronpuchert** [comment]
Doesn't this fit better in `Sema/SemaDeclAttr.cpp`? That's where we're checking the attributes themselves and whether they make sense on a declaration. This would seem like a good place to check against previous declarations. Though to be fair, I don't know how late the late-parsed attributes are parsed, is that really at the end of the file?

The analysis in `Analysis/ThreadSafety.cpp` is based on the CFG of a function definition and at least so far doesn't really care about multiple declarations. But if it's the only way to make it work I guess we'll have to bite the bullet.

**Comment by kevmw** [comment]
I'm the person who asked @tbaederr last year if this could be added for the benefit of QEMU, and as I'm looking at it again now, I thought maybe I should leave a comment here.

In QEMU, we currently handle the problem with a coding convention (public functions get TSA attributes _only_ in the header) and rely on manual code review that the convention is actually obeyed. Obviously, having something automatable like this handled by the compiler would be much nicer and less error prone, so we would still like to see this implemented eventually in some form.

> I could imagine this being useful for RequiresCapability in cases where you only want to require the capability internally as an implementation detail, but not make it part of the public interface of the function.

Not sure what scenario you have in mind there. How is requiring a capability ever something internal? Aren't required capabilities by definition part of the contract of a function?

For example, if a function uses RequiresCapability to specify that it must be called with some lock held and you call it without that lock, that is a bug. If the function takes a lock only internally, then it wouldn't specify that in a TSA attribute because these attributes are only about the interface of the function.

I would actually compare this to function declarations with conflicting calling conventions. This is always a hard compiler error – calling the function with a different calling convention would be a bug, just like calling a function without the right capabilities. (Though admittedly, differing calling conventions would typically result in more obvious bugs if the compiler didn't catch them...)

Am I missing some legitimate scenario in which your code snippet shouldn't cause a warning here because I'm primarily thinking of capabilities as used with locks?

> Yeah, it's a tricky question. On one hand there have been previous issues like #42194 (which this would basically address), and it would definitely improve consistency and prevent beginner's mistakes. On the other hand it seems useful to allow adding attributes e.g. to third-party libraries where you don't control the headers.

I'm not sure that this mistake is one that only beginners would make.

But I agree that adding attributes to functions from third-party libraries is important. I think this is why I first suggested the following in the [downstream tracker](https://issues.redhat.com/browse/RHEL-7269): "I propose that clang should warn if a function definition has TSA attributes, a previous declaration of the same function exists and the attribute isn't present on **at least one** of the previous declarations."

But come to think of it, I'm not sure any more if this really makes a difference – typically, you don't even have a definition for functions from third-party libraries that could trigger the warning, only declarations. The exception are inline functions in headers, but even then, what you get is a declaration with the attribute and a definition without it, which still wouldn't trigger the warning. So do we actually have a problem here?

**Comment by aaronpuchert** [comment]
> > Yeah, it's a tricky question. On one hand there have been previous issues like #42194 (which this would basically address), and it would definitely improve consistency and prevent beginner's mistakes. On the other hand it seems useful to allow adding attributes e.g. to third-party libraries where you don't control the headers.
> 
> I'm not sure that this mistake is one that only beginners would make.

Sure, don't read too much into that statement.

But as a rule of thumb, static analysis attributes always belong on the publicly visible declaration. Otherwise, the caller can't see them. This is true not only for thread safety attributes, but also nullability attributes, handle attributes, type state attributes, and so on. I'm not sure how they're handling it. For `[[noreturn]]` we [have to emit an error](https://eel.is/c++draft/dcl.attr.noreturn#1).

> But I agree that adding attributes to functions from third-party libraries is important. I think this is why I first suggested the following in the [downstream tracker](https://issues.redhat.com/browse/RHEL-7269): "I propose that clang should warn if a function definition has TSA attributes, a previous declaration of the same function exists and the attribute isn't present on **at least one** of the previous declarations."

Ok, that's an interesting twist. I could live with that, but I'm not sure if we handle any other atttributes this way.

> But come to think of it, I'm not sure any more if this really makes a difference – typically, you don't even have a definition for functions from third-party libraries that could trigger the warning, only declarations. The exception are inline functions in headers, but even then, what you get is a declaration with the attribute and a definition without it, which still wouldn't trigger the warning. So do we actually have a problem here?

What I was thinking about was warning on any redeclaration that's adding attributes. Sure, if you only do this on definitions, adding attributes to third-party library wouldn't be an issue.

**Comment by kevmw** [comment]
> But as a rule of thumb, static analysis attributes always belong on the publicly visible declaration. Otherwise, the caller can't see them.

Yes, the theory is obvious and easy to understand, at least once you think of it. Applying it in practice is where I made the experience that mistakes happen a bit too easily.

When you have a mix of static and public functions (so sometimes the annotations are on the definition and sometimes in the header), you're bound to miss a missing "static" sooner or later even if you know the theory. Being careful only gets you so far, so it would be really nice to get that support from the compiler. (And after all, TSA is all about getting support from the compiler instead of trying to be careful yourself.)

Another way to deal with the mistakes would be just outright forbidding most TSA attributes on definitions for public functions, though that's even more restrictive ("TSA attributes on the definition have to be empty" vs. "TSA attributes on the definition have to be a subset of the attributes on declarations").

What I think would make me prefer having the annotations in both places is that when editing a source file, it is slightly inconvenient to have to look at the header to know which assumptions it can make, and it's also slightly inconvenient to update a function in both places if you can't just copy the line, but they have to be different. But that's merely a question of convenience, so if forbidding works better for you, I could live with that, too.

> Ok, that's an interesting twist. I could live with that, but I'm not sure if we handle any other atttributes this way.

Yeah, I'm not sure if there are any that work the exact same way (though I also didn't actively look for one). But as @AaronBallman said above, the attributes are additive, so comparing the definition with the final set of TSA annotations of all declarations combined – which is the same as checking for "at least one" – would seem to make a lot of sense to me.

But as I concluded, checking all previous declarations for function definitions (and only for definitions) should actually work fine in practice, too, unless you're doing crazy things like writing your own implementation with TSA annotations while using a third-party header file for it that doesn't have them. In which case you should probably just copy that header file instead, or locally turn off the warning for that function.

I also compared TSA annotations to calling conventions. There it's possible to have the attribute only on the first declaration and leave it out on later ones (but not the other way around, and obviously also not specifying two different calling conventions). Modelling it after that should work, too, though I'm not sure if the consistency with another attribute is worth the additional limitation.

So there are options and the exact mode is for you guys to decide, I don't mind too much either way.

> What I was thinking about was warning on any redeclaration that's adding attributes. Sure, if you only do this on definitions, adding attributes to third-party library wouldn't be an issue.

Ah, I see. Maybe I misunderstood the code of this PR, but I thought it only checks definitions. And that's honestly all I think we need. It will always be possible to fool TSA with declarations that one file sees and another doesn't, but if we can warn in the common case of a function with one declaration in a header and a definition in a specific source file, it would probably catch 99% of cases in practice.

File: clang/include/clang/Basic/DiagnosticSemaKinds.td [hunk 1]
```diff
@@ -3978,7 +3978,9 @@ def warn_acquired_before : Warning<
 def warn_acquired_before_after_cycle : Warning<
   "cycle in acquired_before/after dependencies, starting with '%0'">,
   InGroup<ThreadSafetyAnalysis>, DefaultIgnore;
-
+def warn_attribute_mismatch : Warning<
+  "attribute mismatch between function definition and declaration of %0">,
+  InGroup<ThreadSafetyAnalysis>, DefaultIgnore;
```
Comment (aaronpuchert) [inline clang/include/clang/Basic/DiagnosticSemaKinds.td:3983]:
```suggestion
  InGroup<ThreadSafetyAttributes>, DefaultIgnore;
```

**Comment by aaronpuchert** [comment]
> Maybe I misunderstood the code of this PR, but I thought it only checks definitions.

That's my understanding as well. I was just thinking about the usual requirement of putting the attribute on the first declaration (like the standard mandates it for `[[noreturn]]`). Users tend the think of _the_ declaration and the definition, but in general there can be an arbitrary number of declarations, at most one of which is a definition.

The idea behind the standard requirement (to my understanding) is that the attribute set doesn't "change" throughout the compilation unit. Of course you have to put any uses after the first declaration, and with that requirement you can analyze uses directly after parsing them without getting a different result from doing it at the end, with all attributes accumulated.

That being said, I'm fine with the idea. I just think this should live in Sema instead of Analysis. That's where we're doing semantic checks of the attributes themselves.

File: clang/lib/Analysis/ThreadSafety.cpp [hunk 2]
```diff
@@ -2260,6 +2260,37 @@ static bool neverReturns(const CFGBlock *B) {
...
+static void diagnoseMismatchedFunctionAttrs(const FunctionDecl *FD,
+                                            ThreadSafetyHandler &Handler) {
+  assert(FD);
+  FD = FD->getDefinition();
+  assert(FD);
+  auto FDArgs = collectAttrArgs<RequiresCapabilityAttr>(FD);
+
+  for (const FunctionDecl *D = FD->getPreviousDecl(); D;
+       D = D->getPreviousDecl()) {
+    auto DArgs = collectAttrArgs<RequiresCapabilityAttr>(D);
```
Comment (aaronpuchert) [inline clang/lib/Analysis/ThreadSafety.cpp:2283]: I don't see why we should restrict this to `requires_capability`. I think pretty much all attributes should be visible to callers except `no_thread_safety_analysis`. This is also [documented](https://clang.llvm.org/docs/ThreadSafetyAnalysis.html):

> Q. Should I put attributes in the header file, or in the .cc/.cpp/.cxx file?
>
> (A) Attributes are part of the formal interface of a function, and should always go in the header, where they are visible to anything that includes the header. Attributes in the .cpp file are not visible outside of the immediate translation unit, which leads to false negatives and false positives.

And later:

> Unlike the other attributes, `NO_THREAD_SAFETY_ANALYSIS` is not part of the interface of a function, and should thus be placed on the function definition (in the `.cc` or `.cpp` file) rather than on the function declaration (in the header).

**Review by aaronpuchert — COMMENTED** [review]
This probably fits better in `Sema/SemaDeclAttr.cpp`. Maybe `checkLockFunAttrCommon` is a good place. Add a test to `clang/test/Sema/attr-capabilities.c{,pp}`.

**Comment by erichkeane** [comment]
Rather than this being "not added in the header file", should we just make this one of the attributes that is disallowed after the thing has been 'referenced'?  Or is that a dumb suggestion here?

**Comment by aaronpuchert** [comment]
That might not be enough. A function might not be used (or even referenced) in the TU that defines it, but only in other TUs. But it would certainly catch a number of issues already.

**Comment by erichkeane** [comment]
> That might not be enough. A function might not be used (or even referenced) in the TU that defines it, but only in other TUs. But it would certainly catch a number of issues already.

Right, though catching that ends up being pretty impossible. The most you could do is determine that the declaration in the defining TU doesn't have it, but not that other forward declarations not visible could catch it.

Unfortunately, unless an attribute contributes to mangling (which, should we consider that?) there is no real way to catch every case.  

Perhaps we could make this a non-inheritable attribute as well, which should catch at least the "added in future declarations".  I think we in a few places we actually DO disallow adding attributes to not-the-first-declaration (which could be done in 'merge', but even just SemaDeclAttr could figure that out), but its not particularly perfect.

So in summary:
-This could/should be caught in Sema, either MergeDeclAttrs or the SemaDeclAttr handling

-We can't catch cross-TU issues here unless we make `RequiresCapability` change mangling.

**Comment by aaronpuchert** [comment]
> > That might not be enough. A function might not be used (or even referenced) in the TU that defines it, but only in other TUs. But it would certainly catch a number of issues already.
> 
> Right, though catching that ends up being pretty impossible. The most you could do is determine that the declaration in the defining TU doesn't have it, but not that other forward declarations not visible could catch it.

Correct, these are all just heuristics. Functions are not typically "forward-declared", so they often have a canonical public declaration. But it is of course possible to declare functions in multiple header files.

> Unfortunately, unless an attribute contributes to mangling (which, should we consider that?) there is no real way to catch every case.

No, I think we don't want mangling. One of the design goals of Thread Safety Analysis was that the attributes should be "erasable", so that they can be dropped for compilers that don't understand them. (See for example https://static.googleusercontent.com/media/research.google.com/en/us/pubs/archive/42958.pdf.)

**Comment by tbaederr** [comment]
Updated the branch.

This now warns:
```c++

#define LOCKABLE            __attribute__ ((lockable))
#define EXCLUSIVE_LOCKS_REQUIRED(...)  __attribute__ ((exclusive_locks_required(__VA_ARGS__)))

struct LOCKABLE Mutex {};
struct Mutex mu1;
struct Mutex mu2;

int Foo_fun1(int i)
  // EXCLUSIVE_LOCKS_REQUIRED(mu2)
  ;

void callsFoo() {
  int a = 10;
  Foo_fun1(a);
}

int Foo_fun1(int i)
  EXCLUSIVE_LOCKS_REQUIRED(mu1)
{
  return i;
}
```
```console
./array.cpp:33:5: warning: attribute mismatch between function declarations of 'Foo_fun1' [-Wthread-safety-attributes]
   33 | int Foo_fun1(int i)
      |     ^
1 warning generated.
```
I started collecting the attribute arguments in a `std::set`, but didn't really end up using them in the end, at least not yet.

**Review by aaronpuchert — COMMENTED** [review]
So the semantics here are: if there is any `requires_capability` attribute on a function, it needs to exactly match the set of `requires_capability` attributes on every previous declaration? Or in other words: the attributes are either inherited or need to be exactly repeated? To me that sounds fine, just wanted to clarify it.

It seems that inheritance results in duplicated attributes, but we eliminate those duplicates in `ThreadSafetyAnalyzer::getMutexIDs` based on `CapabilityExpr::equals`. If we want to allow repetition, we should probably compare in the same way. Which means we probably need to move this back into `ThreadSafety.cpp`, even though it's an attribute warning. (I don't think we want to use Thread Safety internals in Sema.)

**Comment by tbaederr** [comment]
> So the semantics here are: if there is any `requires_capability` attribute on a function, it needs to exactly match the set of `requires_capability` attributes on every previous declaration? Or in other words: the attributes are either inherited or need to be exactly repeated? To me that sounds fine, just wanted to clarify it.

Yes, I think that makes sense. The attributes should match for all (visible) declarations. _If_ the definiton has an attribute too, it should also match.

File: clang/lib/Sema/SemaDeclAttr.cpp [hunk 3]
```diff
@@ -5794,6 +5794,30 @@ static void handleRequiresCapabilityAttr(Sema &S, Decl *D,
...
   D->addAttr(RCA);
+
+  if (const auto *FD = dyn_cast<FunctionDecl>(D)) {
+
+    auto collectExprs = [](const FunctionDecl *FuncDecl) {
+      std::set<const ValueDecl*> Args;
+      for (const auto *A : FuncDecl->specific_attrs<RequiresCapabilityAttr>()) {
+        for (const Expr *E : A->args()) {
+          if (const auto *DRE = dyn_cast<DeclRefExpr>(E))
+            Args.insert(DRE->getDecl());
```
[the remark this answers is not in the document]
Comment (tbaederr) [reply clang/lib/Sema/SemaDeclAttr.cpp:5805]: All the `DeclRefExpr`s are different, so even passing `mu1, mu1` to the attribute results in two different expressions unfortunately.

[the remark this answers is not in the document]
Comment (aaronpuchert) [reply clang/lib/Sema/SemaDeclAttr.cpp:5805]: Yes, in fact most expressions won't be `DeclRefExpr`s but `MemberExpr`s. You get a `DeclRefExpr` when you reference a global variable, but most mutexes are going to be class members. When you reference them, you get a `MemberExpr` with an implicit `CXXThisExpr`. The TIL has facilities for semantically comparing expressions. You can use `SExprBuilder::translateAttrExpr` to translate an expression as `CapabilityExpr` and then compare them using `CapabilityExpr::equals`. This will do the right comparison, e.g. in `mu1, mu1` it will consider both expressions equal.

**Comment by tbaederr** [comment]
Moved it back into `ThreadAnalysis.cpp`. I'm still only comparing the number of args, but I'd like better diagnostics for this. I'm just not sure what the ideal diagnostic output would be.

**Comment by tbaederr** [comment]
This example from the failing test should warn I think:
```c++

class Foo {
public:
  void foo1();
  void foo2();
  void foo3(Foo *other);

  template<class T>
  void fooT1(const T& dummy1);

  template<class T>
  void fooT2(const T& dummy2) EXCLUSIVE_LOCKS_REQUIRED(mu_);

  Mutex mu_;
  int a GUARDED_BY(mu_);
};

void Foo::foo2() EXCLUSIVE_LOCKS_REQUIRED(mu_) {
  a = 2;
}
```
Since the definition of `foo2` specifies attributes that aren't present in the previous declaration.

**Comment by kevmw** [comment]
> This example from the failing test should warn I think
> [...]
> Since the definition of foo2 specifies attributes that aren't present in the previous declaration.

Yes, this looks exactly like the thing we want to protect against to me (because chances are that the declaration is actually in a header file and other source files will check their calls only against this declaration, so real bugs can be expected to result from this).

**Comment by tbaederr** [comment]
Hmm, so this is causing problems:

```c++

namespace DoubleLockBug {
  class Foo {
  public:
    Mutex mu_;
    int a GUARDED_BY(mu_);
    void foo1() EXCLUSIVE_LOCKS_REQUIRED(mu_);
  };
  void Foo::foo1() EXCLUSIVE_LOCKS_REQUIRED(mu_) {
    a = 0;
  }
}
```

The AST is:
```
CXXMethodDecl 0x7d3d96ebf710 parent 0x7d3d96ebf118 prev 0x7d3d96ebf450 <./array.cpp:57:3, line:59:3> line:57:13 foo1 'void ()'
|-CompoundStmt 0x7d3d96ebfa70 <col:50, line:59:3>
| `-BinaryOperator 0x7d3d96ebfa48 <line:58:5, col:9> 'int' lvalue '='
|   |-MemberExpr 0x7d3d96ebf9e8 <col:5> 'int' lvalue ->a 0x7d3d96ebf398
|   | `-CXXThisExpr 0x7d3d96ebf9d0 <col:5> 'DoubleLockBug::Foo *' implicit this
|   `-IntegerLiteral 0x7d3d96ebfa20 <col:9> 'int' 0
|-RequiresCapabilityAttr 0x7d3d96ebf890 <line:15:56, col:92> Inherited exclusive_locks_required
| `-MemberExpr 0x7d3d96ebf5f8 <line:55:42> 'Mutex' lvalue ->mu_ 0x7d3d96ebf328 non_odr_use_unevaluated
|   `-CXXThisExpr 0x7d3d96ebf5e0 <col:42> 'DoubleLockBug::Foo *' implicit this
`-RequiresCapabilityAttr 0x7d3d96ebf930 <line:15:56, col:92> exclusive_locks_required
  `-MemberExpr 0x7d3d96ebf8f8 <line:57:45> 'Mutex' lvalue ->mu_ 0x7d3d96ebf328 non_odr_use_unevaluated
    `-CXXThisExpr 0x7d3d96ebf8e0 <col:45> 'DoubleLockBug::Foo *' implicit this

CXXMethodDecl 0x7d3d96ebf450 <./array.cpp:55:5, line:15:94> line:55:10 foo1 'void ()'
`-RequiresCapabilityAttr 0x7d3d96ebf630 <line:15:56, col:92> exclusive_locks_required
  `-MemberExpr 0x7d3d96ebf5f8 <line:55:42> 'Mutex' lvalue ->mu_ 0x7d3d96ebf328 non_odr_use_unevaluated
    `-CXXThisExpr 0x7d3d96ebf5e0 <col:42> 'DoubleLockBug::Foo *' implicit this
```

So the definition has `mu_` twice, which causes us to diagnose mismatching attributes. Is there any sort of facility to deduplicate capabilities? Declaring a `std::set` of `CapabilityExpr` doesn't seem to work because there's no `operator<` for those.

**Comment by aaronpuchert** [comment]
> Is there any sort of facility to deduplicate capabilities?

The existing code uses `CapExprSet::push_back_nodup`.

**Comment by tbaederr** [comment]
Meh, adding the attribute to the definition and declaration of a member function results in duplicated diagnostics:
https://godbolt.org/z/bcjWca71G

**Comment by aaronpuchert** [comment]
We don't deduplicate "requires" attributes but check them directly. The result is not ideal, but this is slightly more efficient, and the duplicate warning is probably not a big issue. Especially since usually the attributes are not repeated.

File: clang/lib/Analysis/ThreadSafety.cpp [hunk 4]
```diff
@@ -2282,6 +2308,9 @@ void ThreadSafetyAnalyzer::runAnalysis(AnalysisDeclContext &AC) {
   const NamedDecl *D = walker.getDecl();
   CurrentFunction = dyn_cast<FunctionDecl>(D);
 
+  if (isa<FunctionDecl, ObjCMethodDecl>(D))
```
Comment (aaronpuchert) [inline clang/lib/Analysis/ThreadSafety.cpp:2311]: This shouldn't be needed. It should be always true, and even if it wasn't, the check doesn't rely on it.

File: clang/test/SemaCXX/warn-thread-safety-analysis.cpp [hunk 5]
```diff
@@ -2249,7 +2249,7 @@ void fooF1(Foo *f) EXCLUSIVE_LOCKS_REQUIRED(f->mu_) {
   f->a = 1;
 }
 
-void fooF2(Foo *f);
+void fooF2(Foo *f); // expected-warning {{attribute mismatch between function declarations of 'fooF2'}}
 void fooF2(Foo *f) EXCLUSIVE_LOCKS_REQUIRED(f->mu_) {
```
Comment (aaronpuchert) [inline clang/test/SemaCXX/warn-thread-safety-analysis.cpp:2253]: Shouldn't the warning be on the second declaration with a note on the first declaration?

File: clang/lib/Analysis/ThreadSafety.cpp (hunk 4 above)
Comment (tbaederr) [reply clang/lib/Analysis/ThreadSafety.cpp:2311]: The `dyn_cast` right above is confusing then :thinking:

File: clang/test/SemaCXX/warn-thread-safety-analysis.cpp (hunk 5 above)
Comment (tbaederr) [reply clang/test/SemaCXX/warn-thread-safety-analysis.cpp:2253]: I guess that would make a little more sense, but swapping them also causes the note to not get emitted somehow.

Comment (tbaederr) [reply clang/test/SemaCXX/warn-thread-safety-analysis.cpp]: @aaronpuchert Is the current order fine or do you want me to investigate the missing note?

File: clang/lib/Analysis/ThreadSafety.cpp (hunk 4 above)
Comment (aaronpuchert) [reply clang/lib/Analysis/ThreadSafety.cpp]: The way I understand this, `isa<FunctionDecl, ObjCMethodDecl>` checks if the argument is a `FunctionDecl` or `ObjCMethodDecl`, and that is always true. The `dyn_cast` is to `FunctionDecl` specifically.

File: clang/lib/Sema/AnalysisBasedWarnings.cpp [hunk 6]
```diff
@@ -2073,6 +2073,18 @@ class ThreadSafetyReporter : public clang::threadSafety::ThreadSafetyHandler {
...
+  void handleAttributeMismatch(const NamedDecl *ThisDecl,
+                               const NamedDecl *PrevDecl) override {
+    PartialDiagnosticAt Warning(ThisDecl->getLocation(),
+                                S.PDiag(diag::warn_attribute_mismatch)
+                                    << ThisDecl);
+    Warnings.emplace_back(std::move(Warning), getNotes());
+
+    PartialDiagnosticAt Note(PrevDecl->getLocation(),
+                             S.PDiag(diag::note_previous_decl) << PrevDecl);
+    Warnings.emplace_back(std::move(Note), getNotes());
```
Comment (aaronpuchert) [inline clang/lib/Sema/AnalysisBasedWarnings.cpp:2085]: The problem is likely here. You shouldn't add the note as a warning, but as a note to the corresponding warning:
```suggestion
    PartialDiagnosticAt Note(PrevDecl->getLocation(),
                             S.PDiag(diag::note_previous_decl) << PrevDecl);
    Warnings.emplace_back(std::move(Warning), getNotes(Note));
```

Comment (tbaederr) [reply clang/lib/Sema/AnalysisBasedWarnings.cpp]: Ah yes, that makes sense. Thanks!

File: clang/lib/Analysis/ThreadSafety.cpp [hunk 7]
```diff
@@ -2263,6 +2265,27 @@ static bool neverReturns(const CFGBlock *B) {
...
+  };
+
+  auto NDArgs = collectCapabilities(ND);
+  for (const Decl *D = ND->getPreviousDecl(); D; D = D->getPreviousDecl()) {
+    auto DArgs = collectCapabilities(D);
+
+    for (const auto &[A, B] : zip_longest(NDArgs, DArgs)) {
+      if (!A || !B || !A->equals(*B))
+        Handler.handleAttributeMismatch(ND, cast<NamedDecl>(D));
+    }
```
Comment (aaronpuchert) [inline clang/lib/Analysis/ThreadSafety.cpp:2285]: Since we inherit the attributes and deduplicate capabilities here via `push_back_nodup`, this seems to be equivalent to just comparing sizes. We would only have to go through the vectors if we actually wanted to find the missing attribute.

Equivalently, we could also collect the capabilities from attributes on the first declaration, and then on subsequent declarations check _non-inherited_ attributes against the capability set of the first declaration. That would also make it easy to point out which capability is missing on the first declaration. So we can print a better diagnostic.

**Review by aaronpuchert — COMMENTED** [review]
I've thought a bit more about what inheritance means for this and I think it makes our job easier. We can probably collect capabilities on the first declaration and then just check that the non-inherited attributes on subsequent declarations are already contained in that set.

Also, this should be extended to almost all other thread safety attributes. For a start, do one or two so that we can see what it looks like.

**Comment by tbaederr** [comment]
> Also, this should be extended to almost all other thread safety attributes. For a start, do one or two so that we can see what it looks like.

This ends up being pretty ugly since we're collecting the arguments right now, but most of the other attributes don't have arguments or have no `::args()`.

**Comment by aaronpuchert** [comment]
Maybe a single-pass algorithm would be better in the end: we find the first declaration, collect all attributes into a variety of of `CapExprSet`s, and then go through all successor declarations and check their attributes against these sets, emitting a warning if we can't find the capability expression.

We already have such a collection in `handleCall`, and we're going to add another one in #110523. We should probably factor that out, but that's not something I'd ask you to do in this change. You can just duplicate it for now.
